### PR TITLE
[SPARK-23189][Core][Web UI] Reflect stage level blacklisting on executor tab 

### PR DESCRIPTION
**Pull request by attilapiros**
## What changes were proposed in this pull request?

The purpose of this PR to reflect the stage level blacklisting on the executor tab for the currently active stages.

After this change in the executor tab at the Status column one of the following label will be:

- "Blacklisted" when the executor is blacklisted application level (old flag)
- "Dead" when the executor is not Blacklisted and not Active
- "Blacklisted in Stages: [...]" when the executor is Active but the there are active blacklisted stages for the executor. Within the [] coma separated active stageIDs are listed.
- "Active" when the executor is Active and there is no active blacklisted stages for the executor

## How was this patch tested?

Both with unit tests and manually.

#### Manual test

Spark was started as:

```bash
 bin/spark-shell --master "local-cluster[2,1,1024]" --conf "spark.blacklist.enabled=true" --conf "spark.blacklist.stage.maxFailedTasksPerExecutor=1" --conf "spark.blacklist.application.maxFailedTasksPerExecutor=10" 
```

And the job was:
```scala
import org.apache.spark.SparkEnv

val pairs = sc.parallelize(1 to 10000, 10).map { x =>
  if (SparkEnv.get.executorId.toInt == 0) throw new RuntimeException("Bad executor")
  else  {
    Thread.sleep(10)
    (x % 10, x)
  }
}
  
val all = pairs.cogroup(pairs)
  
all.collect()
```

UI screenshots about the running:

- One executor is blacklisted in the two stages: 

One executor is blacklisted in two stages

- One stage completes the other one is still running: 

One stage completes the other is still running

- Both stages are completed: 

Both stages are completed

### Unit tests

In AppStatusListenerSuite.scala both the node blacklisting for a stage and the executor blacklisting for stage are tested.

### CODE REVIEW

**Comment by attilapiros** [comment]
cc @squito @tgravescs

**Comment by tgravescs** [comment]
Jenkins, ok to test

**Comment by tgravescs** [comment]
Jenkins, okay to test

File: core/src/main/resources/org/apache/spark/ui/static/executorspage.js [hunk 1]
```diff
@@ -25,9 +25,13 @@ function getThreadDumpEnabled() {
     return threadDumpEnabled;
 }
 
-function formatStatus(status, type) {
+function formatStatus(status, type, row) {
```
Comment (squito) [inline core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: can you move the

```js
if (row.isBlacklisted) return "Blacklisted";
```
into this function as well?  no reason for it to be separated.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 2]
```diff
@@ -594,12 +606,24 @@ private[spark] class AppStatusListener(
 
       stage.executorSummaries.values.foreach(update(_, now))
       update(stage, now, last = true)
+
+      val executorIdsForStage = stage.executorSummaries.keySet
+      executorIdsForStage.foreach { executorId =>
+        liveExecutors.get(executorId).foreach { exec =>
+          removeBlackListedStageFrom(exec, event.stageInfo.stageId, now)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: I'm just doing a really quick scan here, but I don't understand why changes here are necessary.  You don't get these events for blacklisting within a stage.

or is this a bug in the current code, and something we should fix in the pending 2.3 release?

**Review by squito — CHANGES_REQUESTED** [review]
a few comments from a quick scan

**Comment by ajbozarth** [comment]
I'll take a look at the code when I have a moment, but from a UI perspective only have one issue. Having the status of `Blacklisted in Stages: [...]` when an exec is Active could be easily misunderstood. I think it should still say `Active`, though `Active (Blacklisted in Stages: [...])` is a bit wordy so I'm not sure how best to do so.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 2 above)
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: I guess github diff collapse tricked us here. This changes belongs to the method onStageCompleted (and definitely not for onExecutorUnblacklisted). This is where I remove completed stages from the live executors.

**Comment by attilapiros** [comment]
@ajbozarth What about `Active except for Stages: [...]`?

File: core/src/main/resources/org/apache/spark/ui/static/executorspage.js (hunk 1 above)
Comment (attilapiros) [reply core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: Ok, done.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 2 above)
Comment (squito) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: whoops, my fault!  thanks for explaining

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala [hunk 3]
```diff
@@ -254,6 +255,7 @@ private class LiveExecutor(val executorId: String, _addTime: Long) extends LiveE
   var totalShuffleRead = 0L
   var totalShuffleWrite = 0L
   var isBlacklisted = false
+  var blacklistedInStages: Set[Int] = TreeSet()
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: any particular reason you chose `TreeSet`, and not just `scala.collection.immutable.Set` (scala has a default implementation, specialized for a small number of items, then goes to a `HashSet`) -- see use of `Map` as implementation for `executorLogs`.

**Comment by squito** [comment]
code looks good to me, but lets what @tgravescs @ajbozarth  say.

@ajbozarth it is wordy, but I think `Active (Blacklisted in Stages: [...])` is probably the best of the options so far.

**Comment by ajbozarth** [comment]
I agree that despite being wordy `Active (Blacklisted in Stages: [...])` looks best, could you update it and post screenshots to confirm?

And if @squito says the code looks good I trust him (I'm a bit busy atm to do a proper code review).

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 3 above)
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: I have chosen it  because of the ordering as I assumed the user would like to see here a monoton not decreasing ordering of stageIDs. If it is not needed of course I can change it.

**Comment by attilapiros** [comment]
@ajbozarth Ok. For tomorrow I can update both the label and the screenshots.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 3 above)
Comment (squito) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: yeah thats a good reason, sorry I should have thought of that!

File: core/src/main/resources/org/apache/spark/ui/static/executorspage.js [hunk 4]
```diff
@@ -416,8 +422,7 @@ $(document).ready(function () {
                         },
                         {data: 'hostPort'},
                         {data: 'isActive', render: function (data, type, row) {
-                            if (row.isBlacklisted) return "Blacklisted";
-                            else return formatStatus (data, type);
+                                return formatStatus (data, type, row);
```
Comment (squito) [inline core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: weird indentation here?

Comment (squito) [reply core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: ok nevermind -- now I see the old code didn't seem to follow the convention in the rest of this file.  Not sure about the style guide for js, but I think this might be a bit cleaner if we put each element on its own line, so the nesting is cleaner, like it is here:

https://github.com/apache/spark/blob/master/core/src/main/resources/org/apache/spark/ui/static/executorspage.js#L459-L476

Comment (attilapiros) [reply core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: Ok. Done.

**Review by squito — CHANGES_REQUESTED** [review]
sorry @attilapiros after another read I realized we should avoid making so many updates when a stage completes

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 5]
```diff
@@ -594,12 +606,24 @@ private[spark] class AppStatusListener(
...
+      }
     }
 
     appSummary = new AppSummary(appSummary.numCompletedJobs, appSummary.numCompletedStages + 1)
     kvstore.write(appSummary)
   }
 
+  private def removeBlackListedStageFrom(exec: LiveExecutor, stageId: Int, now: Long) = {
+    exec.blacklistedInStages -= stageId
+    liveUpdate(exec, now)
```
[the remark this answers is not in the document]
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Ok. Modified accordingly and retested manually.

**Comment by ajbozarth** [comment]
Saw the new screenshots, they look good

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 6]
```diff
@@ -228,9 +230,12 @@ private[spark] class AppStatusListener(
...
     // Implicitly blacklist every available executor for the stage associated with this node
     Option(liveStages.get((event.stageId, event.stageAttemptId))).foreach { stage =>
-      liveExecutors.values.filter(_.host == event.hostId).foreach { exec =>
-        setStageBlackListStatus(stage, exec.executorId, now)
+      val executorIds = liveExecutors.values.filter(_.host == event.hostId).map(_.executorId)
+      executorIds.foreach { executorId =>
+        setStageBlackListStatus(stage, executorId, now)
       }
+      stage.blackListedExecutors ++= executorIds
+      maybeUpdate(stage, now)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: if `setStageBlacklistStatus` took a Seq or varags for multiple executorIds, you could push the these other updates down into it.

Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Good idea.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala [hunk 7]
```diff
@@ -423,7 +425,8 @@ private class LiveStage extends LiveEntity {
       newAccumulatorInfos(info.accumulables.values),
       None,
       None,
-      killedSummary)
+      killedSummary,
+      blackListedExecutors)
```
[the remark this answers is not in the document]
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Yes, you are right. Let's have a compact api.

**Comment by attilapiros** [comment]
cc @tgravescs

**Comment by tgravescs** [comment]
UI looks good to me, and taking a quick look at the code it looks fine,  sorry haven't had time to do in depth review,but if @squito is good with it I'm fine with it.

**Comment by squito** [comment]
just a quick note -- I realized I was confused about one part of the inner workings of the history server which I want to confirm before I merge this, but got sick and now have a bit of a backlog.  Will get back to this next week.  Sorry for the delay

**Comment by squito** [comment]
merged to master, thanks everyone
